### ISSUE TITLE
Copy-paste detector of the source code

### ISSUE THREAD

**Issue by vityafx** [rust-lang/rust-clippy#9869]
### What it does

I think it is a good idea to traverse the AST parsed from the source code and try looking for duplicates with some degree of similarity (or rather a full copy) from another place. I have recently had a PR made into one of my projects which had precisely the same copied from a hundred lines of code above, even with the doc comments. If I hadn't been reviewing it well and thoroughly, I'd have missed it. I think it is a good idea to automate this and have such checks done in CI, and I think it is possible to do. Besides all, there are other projects already doing that, for example, [jscpd](https://github.com/kucherenko/jscpd) or [cpd](https://pmd.github.io/latest/pmd_userdocs_cpd.html).

### Lint Name

copy-paste

### Category

correctness, suspicious, style

### Advantage

- Remove the duplicate code either in parts or full code branches (functions, methods, impls, comments, parts of a function body).
- Increase the code quality and readability.
- Possibly allow for early refactoring and/or optimisations made by the copy-paste elimination.

### Drawbacks

- Might be comparatively difficult or performance/time-wise costly to find duplicates.
- Full duplicates are rare but would be helpful to reduce; however, partial code duplication might be linted false-positively. In this case, it would require a lint suppression (if it was enabled for the project).

### Example

```rust
struct A;
#[cfg(target_os = "linux")]
impl A {
    fn do(&self) {
        println!("Hello world!");
    }

    fn linux_specific_function(&self) {
        println!("Hello from Linux!");
    }
}

#[cfg(target_os = "windows")]
impl A {
    fn do(&self) {
        println!("Hello world!");
    }

    fn windows_specific_function(&self) {
        println!("Hello from Windows!");
    }
}
```

Could be written as:

```rust
struct A;

impl A {
    fn do(&self) {
        println!("Hello world!");
    }
}

#[cfg(target_os = "linux")]
impl A {
    fn linux_specific_function(&self) {
        println!("Hello from Linux!");
    }
}

#[cfg(target_os = "windows")]
impl A {
    fn windows_specific_function(&self) {
        println!("Hello from Windows!");
    }
}
```

**Comment by vityafx** [comment]
Partially crosses https://github.com/rust-lang/rust-clippy/issues/2263
